aten/src/ATen/native/cuda/Reduce.cuh

```diff
@@ -508,8 +555,44 @@ static void launch_reduce_kernel(const ReduceConfig& config, const R& reduction)
   AT_CUDA_CHECK(cudaGetLastError());
 }
 
+struct AccumulationBuffer {
+  AccumulationBuffer() {}
+
+  AccumulationBuffer(size_t acc_t_size, size_t out_t_size, char* out_ptr, int64_t size) {
+    out_ptr_ = (char*)out_ptr;
+    if (out_t_size > acc_t_size) {
+      // reusing output buffer for accumulation.
+      acc_ptr_ = (char*)out_ptr;
+      numerator_ = 1;
+      denominator_ = 1;
+    } else {
+      auto& allocator = *at::globalContext().getTHCState()->cudaDeviceAllocator;
+      buffer_ = allocator.allocate(size);
+      acc_ptr_ = (char*)buffer_.get();
+      numerator_ = acc_t_size;
+      denominator_ = out_t_size;
+      reduce_fraction(numerator_, denominator_);
+    }
+  }
+
+  char* get_acc_slice(char* out_ptr) {
+    if (numerator_ == -1 || acc_ptr_ == nullptr) {
+      return nullptr;
+    }
+    return acc_ptr_ + ((out_ptr - out_ptr_) * numerator_ / denominator_);
+  }
+
+  char* acc_ptr_ = nullptr;
+  char* out_ptr_ = nullptr;
+  float size_factor_ = -1;
+  size_t numerator_ = -1;
+  size_t denominator_ = -1;
+  at::DataPtr buffer_;
+};
+
 template <typename scalar_t, typename out_scalar_t, typename ops_t, typename ident_t=double>
-inline void gpu_reduce_kernel(TensorIterator& iter, const ops_t& ops, ident_t ident=0) {
+inline void gpu_reduce_kernel(TensorIterator& iter, const ops_t& ops, ident_t ident=0,
+                              std::shared_ptr<AccumulationBuffer> acc_buf_ptr={}) {
```

Comment:
Does `acc_buf_ptr` really need to be reference-counted/shared?

Why can't we just pass around a raw `AccumulationBuffer *` pointer, and allocate the `AccumulationBuffer` on the stack in the top-level call (so it will be properly deallocated when that call returns)

Reply:
Fair enough, I'll switch to unique_ptr and pass raw pointer instead.